libcaf_core/test/ipv4_endpoint.cpp

```diff
@@ -0,0 +1,142 @@
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright 2011-2019 Dominik Charousset                                     *
+ *                                                                            *
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#define CAF_SUITE ipv4_endpoint
+
+#include "caf/ipv4_endpoint.hpp"
+
+#include "caf/test/unit_test.hpp"
+
+#include <vector>
+
+#include "caf/actor_system.hpp"
+#include "caf/actor_system_config.hpp"
+#include "caf/binary_deserializer.hpp"
+#include "caf/byte.hpp"
+#include "caf/ipv4_address.hpp"
+#include "caf/serializer_impl.hpp"
+#include "caf/span.hpp"
+
+using namespace caf;
+
+namespace {
+
+struct test_data {
+  using endpoint = ipv4_endpoint;
+  using comparison_testcase = std::pair<endpoint, endpoint>;
+  using to_string_testcase = std::pair<endpoint, std::string>;
+
+  test_data() {
+    const auto addr = make_ipv4_address;
+    // different ip but same port
+    add({addr(127, 0, 0, 1), 8888}, {addr(127, 0, 0, 2), 8888});
+    add({addr(192, 168, 178, 1), 8888}, {addr(245, 114, 2, 89), 8888});
+    add({addr(188, 56, 23, 97), 1211}, {addr(189, 22, 36, 0), 1211});
+    add({addr(0, 0, 0, 0), 8888}, {addr(255, 255, 255, 1), 8888});
+    // same ip but different port
+    add({addr(127, 0, 0, 1), 111}, {addr(127, 0, 0, 1), 8888});
+    add({addr(192, 168, 178, 1), 8888}, {addr(245, 114, 2, 89), 8888});
+    add({addr(127, 0, 0, 1), 111}, {addr(127, 0, 0, 1), 8888});
+    add({addr(123, 123, 123, 123), 8888}, {addr(123, 123, 123, 123), 8889});
+
+    add({addr(127, 0, 0, 1), 8888}, "127.0.0.1:8888");
+    add({addr(192, 168, 178, 1), 8888}, "192.168.178.1:8888");
+    add({addr(127, 0, 0, 1), 111}, "127.0.0.1:111");
+    add({addr(192, 168, 178, 1), 8888}, "192.168.178.1:8888");
+    add({addr(127, 0, 0, 1), 111}, "127.0.0.1:111");
+    add({addr(123, 123, 123, 123), 8888}, "123.123.123.123:8888");
+  }
+
+  // First endpoint is always smaller than the second.
+  std::vector<comparison_testcase> comparison_testdata;
+  std::vector<to_string_testcase> to_string_testdata;
+
+private:
+  void add(endpoint ep1, endpoint ep2) {
+    comparison_testdata.emplace_back(comparison_testcase{ep1, ep2});
+  }
+
+  void add(endpoint ep, const std::string& str) {
```

Comment:
Having two functions named `add` that write into different vectors is hard to follow. Ideally, we simply initialize the vectors with the data instead of filling them piece by piece. What do you think of this strawman to make the code a bite more readable?


```c++
namespace {

ipv4_endpoint operator"" _ep(const char* str,size_t) {
  std::array<uint8_t, 4> bytes;
  char* next = nullptr;
  bytes[0] = static_cast<uint8_t>(strtol(str, &next, 10));
  for (size_t n = 1; n < 4; ++n) {
    assert(*next == '.');
    str = next + 1;
    bytes[n] = static_cast<uint8_t>(strtol(str, &next, 10));
  }
  assert(*next == ':');
  auto port = static_cast<uint16_t>(strtol(next + 1, nullptr, 10));
  return ipv4_endpoint{ipv4_address{bytes}, port};
}

class fixture {
  actor_system_config cfg;
  actor_system sys{cfg};
  std::vector<std::pair<ipv4_endpoint, ipv4_endpiont>> xs{
    {"127.0.0.1:8888"_ep, "127.0.0.2:8888"_ep},
    // ...
  };
};

#define CHECK_TO_STRING(addr)                                                  \
  CAF_CHECK_EQUAL(to_string(addr), addr##_ep)

} // namespace
```

This user-defined literal is of course not "production ready", but it's good enough as a unit test DSL. Now, you can also write a test for `to_string` simply like this:

```
CAF_TEST(to string) {
  CHECK_TO_STRING("127.0.0.1:8080");
  // ...
}
```

This is what the IPv6 version could look like:

```c++
ipv6_endpoint operator"" _ep(const char* str,size_t) {
  union {
    std::array<uint8_t, 16> xs_bytes;
    std::array<uint16_t, 8> xs;
  };
  union {
    std::array<uint8_t, 16> ys_bytes;
    std::array<uint16_t, 8> ys;
  };
  memset(xs_bytes.data(), 0, xs_bytes.size());
  memset(ys_bytes.data(), 0, xs_bytes.size());
  assert(*str == '[');
  ++str;
  char* next = nullptr;
  auto pull = [&] {
    auto x = static_cast<uint16_t>(strtol(str, &next, 16));
    return detail::to_network_order(x);
  };
  size_t n = 0;
  do {
    xs[n++] = pull();
    assert(next != nullptr);
    str = next + 1;
  } while (*str != ':' && *str != ']');
  if (*str == ':') {
    ++str;
    do {
      ys[0] = pull();
      std::rotate(ys.begin(), ys.begin() + 1, ys.end());
      assert(next != nullptr);
      str = next + 1;
    } while (*next != ']');
  }
  assert(*next == ']');
  assert(*str == ':');
  ++str;
  auto port = static_cast<uint16_t>(strtol(str, nullptr, 10));
  std::array<uint8_t, 16> bytes;
  for (size_t i = 0; i < 16; ++i)
    bytes[i] = xs_bytes[i] | ys_bytes[i];
  return ipv6_endpoint{ipv6_address{bytes}, port};
}
```

Btw, PR #882 adds new `parse` functions. Once that got merged, we can streamline the user-defined literals (but I think they're OK for now to make the unit tests more comprehensible).